### DISCUSSION TITLE
Experimental Multiprocessing

### DISCUSSION THREAD

**Discussion by capellancitizen** [inkstitch/inkstitch#2991] [Ideas]
Main notes:
- Most of the embroidery work happens in Python, so if we want to get any real parallel processing speedup we're probably going to use the `multiprocessing` library to sidestep the GIL. I've done so in the `capellan-multiprocessing` branch, but note that it's still a WIP - I bet the frozen build won't work out of the box at the moment because it doesn't call `freeze_support()` for instance...
- Many elements can have their stitch plans generated in parallel, but there are certain elements that are dependent on the last stitch made by the previous element (specifically, fills without a defined fill start command, but those are probably most fills). In my branch, I break down the list of elements into a list of element "chains" (where the first element in the chain is not dependent on any other element, and each following element is dependent on the element before it). This does mean that the gains we can get from parallelism are somewhat impacted by the ordering of the elements, unless the user places `start` markers for us.
- The main hurdle here is that `inkex` elements (and the `lxml` elements they inherit from) can't be pickled to send to the worker subprocesses, and so neither can `EmbroideryElement`s because they have these as members. To work around this, the branch sends each worker the full svg document and has them parse it themselves so they have a copy to work off of. Because I can't pass elements around, I instead pass around tuples of the xpath to the element and the element class, so the worker threads can reconstruct them and call their methods referencing their local copies of the svg document. This has the benefit of working with much of our code as-is (indeed, the only changes I've made to `EmbroideryElement` were some additional timing logging and some more type information), but it is perhaps a bit inelegant. That said, it doesn't seem to take very long to send the document and parse it, less than a tenth of a second even for a pretty large document.
  - The alternative is probably decoupling the `EmbroideryElement` class from the element it "wraps", by e.g. copying out the attributes it needs (and computing style, composed transform, etc) for embroidery at the start, but that's a big refactor. (It might be convenient because then `EmbroideryElement` doesn't have to be tied to the lifetime of the elements they are based on, which might make managign clones easier. Speaking of clones...)
- `Clone`s currently basically don't benefit from parallellism at all at the moment, because inside `Clone` there's the same sort of element-to-stitch-plan loop as in the `BaseExtension` class. Probably going to have to make `Clone` special such that clones are resolved into their `EmbroideryElement`s at the start before the `EmbroideryElement`s are turned into stitch plans.

And for all that, we get some performance gains, sometimes. With 4 cpu cores (and 4 worker processes), running simple stitch plan generation on some files from the website (plus one other one):
- Inkstitch-Contour-Fill-Sunflower.svg goes from ~1.4s to ~0.7s. Despite all of the fills, they are broken up by other objects so it breaks down into 8 chains and benefits from the parallelism quite well.
- SoftwareFreedomDayFilledSun.svg goes from ~4.3s to ~2.9s. Almost every element is independently embroiderable for this one, so it works fairly well.
- ripple-iris.svg: ~1.2 to ~0.7s
- paon.svg: ~9.8s to ~3.8s (!!)
- Grumpycat.svg gets no noticeable performance improvement, ~12s before and after. For this file, almost all of the time is spent calculating the plan for the large "fuzzy" fill and all of the other elements take less than a second to embroider altogether, so the performance improvement is a wash.
- 22_Kielux_12cm.svg: 50.7s -> 36.8s

These results are honestly a little disappointing, given even with a lot of independently embroiderable elements it doesn't get the boost you'd hope for embroidering up to 4 elements at once. Multi-process profiling is a little annoying so I haven't dug in too deeply yet to figure out what's going on on the worker processes and what's taking up the time. 

Others are of course free to hack on this as they want, I'll try to get back to it later.

**Comment by capellancitizen** [comment]
[Having searched more thoroughly through the issue list I see Lexelby was looking into this approximately a year ago.](https://github.com/inkstitch/inkstitch/issues/836#issuecomment-1657340702) Basically everything they've said in that thread, I can concur. I hacked this branch so we can get a one-click performance comparison, by running threading plan both single nad multithreaded, and timing/logging both. Unfortunately on Windows the naive first attempt has abysmal performance, probably owing to multiprocessing on Windows not having `fork`-type instantiation.

On Windows, with caching off:
For ripple-iris, single-threaded takes ~1s, multi-threaded execution takes ~3.7s.
For Bfly-FSL, single-threaded takes ~3.3s, multi-threaded execution takes 17.7s(!!)

I don't have a Mac to test on, but I suspect that it'll have the same performance issues given it also uses `spawn` and not `fork` by default. Multiprocessing probably (hopefully) isn't a complete dead end, but there's going to be more work than just using the Pool probably.

**Comment by lexelby** [comment]
Thanks for looking at this!  I gotta admit, I got pretty frustrated toward the end.

I guess one thing that didn't occur to be is that we could simply say that multi-core performance is only available on platforms with fork().  I wouldn't mind if only Linux and Mac got a speed boost since people can always use virtualization or dual boot.

**Comment by lexelby** [comment]
Apparently WSL _does_ support fork().  I wonder if there's some way for us to use WSL...

**Reply by capellancitizen** [reply]
According to the docs, the OSX implementation of multiprocessing uses `spawn` by default rather than `fork`, and it looks like they're moving away from `fork` being the default method in 3.14. Possibly this is related to fork having some issues with multithreaded applications and frozen applications (`frozen_support` aside?). Now that I think about it, the simulator is multithreaded, so I wonder if this breaks there, haven't tested it yet.

WSL runs Linux in a VM, doesn't it? Not sure if it's something a Windows application can really benefit from.

We're probably better off looking at other avenues for a speedup. It's a shame because we should be parallelizable, but Python doesn't make that easy.

**Reply by lexelby** [reply]
I think WSL 1 isn't quite virtualization, and WSL 2 is but not in the classic sense.

And yeah, that's the impression I got from my investigation too.  It's just too much work for not enough gain.

I do think there's plenty of room to speed things up, though.  In a lot of cases, ice traded speed for readability, or I've avoided premature optimization -- and maybe now it's time to revisit and do that optimization :)

**Reply by capellancitizen** [reply]
It is unfortunate that we have a problem that can be parallelized, but the language we're using is in the way. I'm not necessarily saying "rewrite stitch planning in Rust or some other language", there's probably other places to look first...

I guess, in practical use the lack of parallelism isn't too big a hit: After the first stitch planning, you'll mostly be editing elements with the Params window, and in doing that you will have already generated entries in the cache for these elements that will be fetched for later stitch plans. Parallelism's main win is whenever the program entirely misses the cache.

**Comment by lexelby** [comment]
I suppose one last possible way to do it would be to spawn multiple processes of Ink/Stitch _at the start_, before we import anything in lib.  One process would be the main one, and the child processes would sit and wait for requests to render portions of the stitch plan.  That way we're doing that 1+ second of launch stuff in parallel, and every process has already parsed the scg and instantiated EmbroideryElement objects for everything.  

What do you think?  Still sounds very complicated to me...

**Reply by capellancitizen** [reply]
If we did that we'd probably need a flag or class for "extensions will want to generate a stitch plan and thus need to spin up the process pool". I guess we at least have a test bed to play with if we want to try it. For Windows in particular I wonder if starting early would even help given how long it takes to clone. This all makes me wonder about the actual details of multiprocessing/pool, just how much does the `spawn` method instantiate upfront? I'd hope it's not the whole code all at once... Maybe this is all just proof the complexity of `multiprocessing` makes it more effort than its worth? Feh.
